File: examples/build-reduced-model.cpp
// ============================================================
// 降阶模型（Reduced Model）：锁定指定关节，构建自由度更少的模型
// 应用场景：
//   - 固定手臂，只研究腿部动力学（减少计算量）
//   - 固定上半身，研究下肢平衡控制
//   - 将复杂机器人拆分为子系统分别优化
//
// 两种构建方式：
//   1. 指定"要锁定的关节"列表（直接指定锁定哪些）
//   2. 指定"要保留的关节"列表（间接：锁定所有其他关节）
// ============================================================

#include "pinocchio/parsers/urdf.hpp"

#include "pinocchio/algorithm/joint-configuration.hpp"
#include "pinocchio/algorithm/model.hpp"  // buildReducedModel

#include <iostream>
#include <algorithm>

// PINOCCHIO_MODEL_DIR is defined by the CMake but you can define your own directory here.
#ifndef PINOCCHIO_MODEL_DIR
  #define PINOCCHIO_MODEL_DIR "path_to_the_model_dir"
#endif

// 辅助函数：检查元素是否在 vector 中（C++11 无 contains）
template<typename T>
bool is_in_vector(const std::vector<T> & vector, const T & elt)
{
  return vector.end() != std::find(vector.begin(), vector.end(), elt);
}

int main(int argc, char ** argv)
{
  using namespace pinocchio;

  const std::string urdf_filename =
    (argc <= 1) ? PINOCCHIO_MODEL_DIR
                    + std::string("/example-robot-data/robots/ur_description/urdf/ur5_robot.urdf")
                : argv[1];

  // Load the urdf model（固定基 UR5，6 DOF）
  Model model;
  pinocchio::urdf::buildModel(urdf_filename, model);

  // ====================================================
  // 方式一：指定"要锁定的关节"列表
  // ====================================================
  std::vector<std::string> list_of_joints_to_lock_by_name;
  list_of_joints_to_lock_by_name.push_back("elbow_joint");
  list_of_joints_to_lock_by_name.push_back("wrist_3_joint"); // 顺序无关（内部会排序）
  list_of_joints_to_lock_by_name.push_back("wrist_2_joint");
  list_of_joints_to_lock_by_name.push_back("blabla"); // 不存在的关节名（会被跳过）

  // 将关节名转为 ID（跳过不存在的名称）
  std::vector<JointIndex> list_of_joints_to_lock_by_id;
  for (std::vector<std::string>::const_iterator it = list_of_joints_to_lock_by_name.begin();
       it != list_of_joints_to_lock_by_name.end(); ++it)
  {
    const std::string & joint_name = *it;
    if (model.existJointName(joint_name))  // 验证关节名是否存在
      list_of_joints_to_lock_by_id.push_back(model.getJointId(joint_name));
    else
      std::cout << "joint: " << joint_name << " does not belong to the model" << std::endl;
  }

  // q_rand：锁定关节将被固定在该配置对应的角度
  // 被锁定关节的运动学参数被"烘焙"到模型几何结构中（变为固定偏置）
  Eigen::VectorXd q_rand    = randomConfiguration(model);
  Eigen::VectorXd q_neutral = neutral(model);
  PINOCCHIO_UNUSED_VARIABLE(q_neutral);  // 仅用于演示 neutral() 用法

  std::cout << "\n\nFIRST CASE: BUILD A REDUCED MODEL FROM A LIST OF JOINT TO LOCK" << std::endl;

  // buildReducedModel(model, joints_to_lock, reference_q)
  // 返回降阶模型（njoints 减少 = 锁定关节数，nq/nv 相应减小）
  // reference_q：被锁定关节固定在此配置
  Model reduced_model = buildReducedModel(model, list_of_joints_to_lock_by_id, q_rand);

  // Print the list of joints in the original model
  std::cout << "List of joints in the original model:" << std::endl;
  for (JointIndex joint_id = 1; joint_id < model.joints.size(); ++joint_id)
    std::cout << "\t- " << model.names[joint_id] << std::endl;

  // Print the list of joints in the reduced model
  std::cout << "List of joints in the reduced model:" << std::endl;
  for (JointIndex joint_id = 1; joint_id < reduced_model.joints.size(); ++joint_id)
    std::cout << "\t- " << reduced_model.names[joint_id] << std::endl;

  // ====================================================
  // 方式二：指定"要保留的关节"列表（等价但更直观）
  // 适用场景：已知要保留哪些关节（如只保留腿部关节）
  // ====================================================
  std::cout << "\n\nSECOND CASE: BUILD A REDUCED MODEL FROM A LIST OF JOINT TO KEEP UNLOCKED"
            << std::endl;

  std::vector<std::string> list_of_joints_to_keep_unlocked_by_name;
  list_of_joints_to_keep_unlocked_by_name.push_back("shoulder_pan_joint");
  list_of_joints_to_keep_unlocked_by_name.push_back("shoulder_lift_joint");
  list_of_joints_to_keep_unlocked_by_name.push_back("wrist_1_joint");

  std::vector<JointIndex> list_of_joints_to_keep_unlocked_by_id;
  for (std::vector<std::string>::const_iterator it =
         list_of_joints_to_keep_unlocked_by_name.begin();
       it != list_of_joints_to_keep_unlocked_by_name.end(); ++it)
  {
    const std::string & joint_name = *it;
    if (model.existJointName(joint_name))
      list_of_joints_to_keep_unlocked_by_id.push_back(model.getJointId(joint_name));
    else
      std::cout << "joint: " << joint_name << " does not belong to the model";
  }

  // 转换：将"保留列表"取补集得到"锁定列表"
  // 遍历所有关节（从 1 开始，0=宇宙关节），不在保留列表中的即锁定
  list_of_joints_to_lock_by_id.clear();
  for (JointIndex joint_id = 1; joint_id < model.joints.size(); ++joint_id)
  {
    const std::string joint_name = model.names[joint_id];
    if (is_in_vector(list_of_joints_to_keep_unlocked_by_name, joint_name))
      continue;   // 保留的关节，跳过
    else
      list_of_joints_to_lock_by_id.push_back(joint_id);  // 其余的锁定
  }

  // 构建第二个降阶模型（只保留 shoulder_pan, shoulder_lift, wrist_1）
  Model reduced_model2 = buildReducedModel(model, list_of_joints_to_lock_by_id, q_rand);

  // Print the list of joints in the second reduced model
  std::cout << "List of joints in the second reduced model:" << std::endl;
  for (JointIndex joint_id = 1; joint_id < reduced_model2.joints.size(); ++joint_id)
    std::cout << "\t- " << reduced_model2.names[joint_id] << std::endl;
}

File: examples/build-reduced-model.py
from pathlib import Path

import numpy as np
import pinocchio as pin

# ============================================================
# 降阶模型：将指定关节锁定在参考配置处，构建自由度更少的模型
# 应用场景：
#   - 人形上身固定，只研究腿部动力学
#   - 手臂固定，研究躯干+腿部的平衡控制
#   - 加速仿真（减少计算量）
# ============================================================

pinocchio_model_dir = Path(__file__).parent.parent / "models"
model_path = pinocchio_model_dir / "example-robot-data/robots"
mesh_dir   = pinocchio_model_dir
urdf_filename = model_path / "ur_description/urdf/ur5_robot.urdf"

# 同时加载运动学模型、碰撞模型、视觉模型（降阶时可一并处理）
model, collision_model, visual_model = pin.buildModelsFromUrdf(urdf_filename, mesh_dir)

print("standard model: dim=" + str(len(model.joints)))
for jn in model.joints:
    print(jn)
print("-" * 30)

# 要锁定的关节名称列表（这些关节在降阶模型中被固定）
# 实际使用时替换为需要固定的关节名
jointsToLock = ["wrist_1_joint", "wrist_2_joint", "wrist_3_joint"]

# 将关节名转为 ID（不存在的名称会被警告跳过）
jointsToLockIDs = []
for jn in jointsToLock:
    if model.existJointName(jn):
        jointsToLockIDs.append(model.getJointId(jn))
    else:
        print("Warning: joint " + str(jn) + " does not belong to the model!")

# 锁定关节时使用的参考配置（被锁定关节固定在此角度）
# 长度必须等于完整模型的 nq
initialJointConfig = np.array([
    0, 0, 0,   # 肩部和肘部关节（保持零位）
    1, 1, 1,   # 被锁定的腕部关节（固定在 1 rad）
])

# ---- Option 1：仅降阶运动学/动力学模型（不含几何模型）----
# 最轻量，适合纯动力学计算（不需要可视化时）
model_reduced = pin.buildReducedModel(model, jointsToLockIDs, initialJointConfig)

# ---- Option 2：同时降阶视觉几何模型（用于可视化）----
# 返回降阶后的运动学模型 + 对应的视觉模型
model_reduced, visual_model_reduced = pin.buildReducedModel(
    model, visual_model, jointsToLockIDs, initialJointConfig
)

# ---- Option 3：同时降阶多个几何模型（视觉 + 碰撞）----
# 返回顺序与传入的 geom_models 列表一致
geom_models = [visual_model, collision_model]
model_reduced, geometric_models_reduced = pin.buildReducedModel(
    model,
    list_of_geom_models=geom_models,
    list_of_joints_to_lock=jointsToLockIDs,
    reference_configuration=initialJointConfig,
)
# 按顺序解包（与传入顺序一致）
visual_model_reduced, collision_model_reduced = (
    geometric_models_reduced[0],
    geometric_models_reduced[1],
)

print("joints to lock (only ids):", jointsToLockIDs)
print("reduced model: dim=" + str(len(model_reduced.joints)))
print("-" * 30)

# ---- Option 4：使用 RobotWrapper（可混用关节名和关节 ID）----
# RobotWrapper 封装了 model + collision_model + visual_model，接口更高层
# reference_configuration 可选，不传则使用零位
mixed_jointsToLockIDs = [jointsToLockIDs[0], "wrist_2_joint", "wrist_3_joint"]
robot = pin.RobotWrapper.BuildFromURDF(urdf_filename, mesh_dir)
reduced_robot = robot.buildReducedRobot(
    list_of_joints_to_lock=mixed_jointsToLockIDs,
    reference_configuration=initialJointConfig,
)

print("mixed joints to lock (names and ids):", mixed_jointsToLockIDs)
print("RobotWrapper reduced model: dim=" + str(len(reduced_robot.model.joints)))
for jn in robot.model.joints:
    print(jn)
